azure-storage-queue/src/main/scala/akka/stream/alpakka/azure/storagequeue/AzureQueueSourceStage.scala

```diff
@@ -4,86 +4,85 @@
+ */
 final case class AzureQueueSourceSettings(
-    waitTimeSeconds: Int,
-    maxBufferSize: Int,
-    maxBatchSize: Int
-) {
-  require(maxBatchSize <= maxBufferSize, "maxBatchSize must be lower or equal than maxBufferSize")
-}
+    initialVisibilityTimeout: Int,
+    batchSize: Int,
+    retrieveRetryTimeout: Option[FiniteDuration]
+)
 
 object AzureQueueSourceSettings {
-  val default = AzureQueueSourceSettings(20, 100, 10)
+  // default initialVisibilityTimeout (30) is from
+  // com.microsoft.azure.storage.queue.QueueConstants.DEFAULT_VISIBILITY_MESSAGE_TIMEOUT_IN_SECONDS
+  val default = AzureQueueSourceSettings(30, 10, None)
 
   /**
-   *  Java API
+   *  Java API: Constructor for [[AzureQueueSourceSettings]]
+   *  @param retrieveRetryTimeout in seconds. If <= 0 retrying of message retrieval is disable.
    */
-  def create(waitTimeSeconds: Int, maxBufferSize: Int, maxBatchSize: Int): AzureQueueSourceSettings =
-    AzureQueueSourceSettings(waitTimeSeconds, maxBufferSize, maxBatchSize)
+  def create(initialVisibilityTimeout: Int, batchSize: Int, retrieveRetryTimeout: Int): AzureQueueSourceSettings = {
+    val rrTimeout = if (retrieveRetryTimeout > 0) Some(retrieveRetryTimeout.seconds) else None
+    AzureQueueSourceSettings(initialVisibilityTimeout, batchSize, rrTimeout)
+  }
 }
 
 private[storagequeue] final class AzureQueueSourceStage(cloudQueue: CloudQueue, settings: AzureQueueSourceSettings)
     extends GraphStage[SourceShape[CloudQueueMessage]] {
   val out: Outlet[CloudQueueMessage] = Outlet("AzureCloudQueue.out")
   override val shape: SourceShape[CloudQueueMessage] = SourceShape(out)
 
-  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) {
-    val buffer = new Queue[CloudQueueMessage]
-
-    private val maxConcurrency = settings.maxBufferSize / settings.maxBatchSize
-    private var maxCurrentConcurrency = maxConcurrency
-    private var currentRequests = 0
-
-    private val successCallback = getAsyncCallback(handleSuccess)
-    private val failureCallback = getAsyncCallback(handleFailure)
-
-    private def canReceiveNewMessages: Boolean = {
-      val currentFreeRequests = (settings.maxBufferSize - buffer.size) / settings.maxBatchSize
-      currentFreeRequests > currentRequests && maxCurrentConcurrency > currentRequests
-    }
-
-    def receiveMessages(): Unit =
-      if (canReceiveNewMessages) {
-        implicit val executionContext = materializer.executionContext
-        currentRequests += 1
-
-        Future {
-          import scala.collection.JavaConverters._
-          val res = cloudQueue.retrieveMessages(settings.maxBatchSize, settings.waitTimeSeconds, null, null)
-          res.asScala.toList
-        } onComplete {
-          case Success(res) => successCallback.invoke(res)
-          case Failure(t) => failureCallback.invoke(t)
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new TimerGraphStageLogic(shape) {
```

Comment:
Alright, this simplified the stage quite a bit. Now it needs to run on a dispatcher for blocking stuff, you can look at https://github.com/akka/alpakka/blob/20ee323a22d5cf8e0ac53948d6b9f323a4b832ae/ftp/src/main/scala/akka/stream/alpakka/ftp/impl/FtpIOGraphStage.scala#L28 for an example how to do that through the initial attributes of the stage